Add Lazy Relay Stats

jcorbin opened the pull request:
r @kriskowal @Raynos

Raynos commented on relay_handler.js:
```diff
@@ -372,6 +392,78 @@ function handleFrameLazily(frame) {
+        )
+    ));
+
+    self.channel.emitFastStat(self.channel.buildStat(
```
TODO: No garantuee that we send the frame out succesfully. frame write could fail.

jcorbin replied:
Not a thing that the old path guaranteed either; same reply applies to all subsequent instances.

Raynos commented on relay_handler.js:
```diff
@@ -372,6 +392,78 @@ function handleFrameLazily(frame) {
+        )
+    ));
+
+    self.channel.emitFastStat(self.channel.buildStat(
```
TODO: No garantuee that we send the frame out succesfully. frame write could fail.

Raynos commented on relay_handler.js:
```diff
@@ -372,6 +392,78 @@ function handleFrameLazily(frame) {
+        )
+    ));
+
+    self.channel.emitFastStat(self.channel.buildStat(
```
TODO: No garantuee that we send the frame out succesfully. frame write could fail.

Raynos commented on relay_handler.js:
```diff
@@ -66,6 +66,17 @@ RelayHandler.prototype.handleLazily = function handleLazily(conn, reqFrame) {
     }
 
+    self.channel.emitFastStat(self.channel.buildStat(
+        'tchannel.inbound.calls.recvd',
```
This must be for CallReq only; not CallReqCont

Raynos replied:
Ok i get it. maybe rename `handleLazily` to `handleLazyCallReq`

jcorbin replied:
fair point, should be renamed, in another pr perhaps

Raynos commented on relay_handler.js:
```diff
@@ -299,6 +310,13 @@ function onIdentified(err) {
+    self.channel.emitFastStat(self.channel.buildStat(
+        'tchannel.relay.latency',
+        'timing',
+        self.outreq.start - self.start,
```
We want `Date.now() - self.start`

jcorbin replied:
good point, will do

Raynos commented on relay_handler.js:
```diff
@@ -358,6 +376,8 @@ function handleFrameLazily(frame) {
     // - v2.Types.CallRequestCont
     var self = this;
 
+    var now = self.channel.timers.now();
```
I'm dubious of this `now` pattern.

I feel like `self.start` / `self.end` etc is better pattern.

jcorbin replied:
we don't have self.end on either of the lazy reqs; I'm already unhappy with how many fields they've gained.

Raynos commented on relay_handler.js:
```diff
@@ -421,6 +513,35 @@ LazyRelayOutReq.prototype.emitError =
+        )
+    ));
+
+    self.channel.emitFastStat(self.channel.buildStat(
```
TODO: double check this is never an error frame ?

jcorbin replied:
It's not, this entire code path (`LazyRelay{In,Out}Req`) will never normally see an error frame error object, the two normal callers are:
- LazyRealyOutReq#onTimeout
- TChannelConnection#resetAll

The only other call is TChannelConnection#onCallErrorFrame, this could only happen during a transition time (when we flip off lazy handling e.g.).  I'd instead rather add a branch to this function that makes it call `req.handleLazily` if such function exists, or similar, thoughts?

Raynos commented on relay_handler.js:
```diff
@@ -464,12 +583,32 @@ function handleFrameLazily(frame) {
+    self.channel.emitFastStat(self.channel.buildStat(
+        'tchannel.inbound.calls.latency',
+        'timing',
+        now - self.inreq.start,
```
This looks like cheating to me. We should include the time for `socket.write()` in this.

`Date.now()`

jcorbin replied:
will do: will move down the now call until after the socket write until right before `_observe*` call tree

Raynos commented on relay_handler.js:
```diff
@@ -498,6 +663,144 @@ function logForwardedError(errFrame) {
+    self.channel.emitFastStat(self.channel.buildStat(
+        'tchannel.inbound.calls.latency',
+        'timing',
+        now - self.inreq.start,
```
This is once more cheating. Need to do a `Date.now()` here to include `socket.write()` time.

jcorbin replied:
same as last, will fix

Raynos commented:
Finished review.

Need a second opinion from @rf

jcorbin commented:
Addressed feedback; the socket write "guarantee" stuff sounds interesting, but it's not a feature the old path had. It may be easier to get out of this new path tho, but I feel should be deferred to future work.

Raynos commented:
Lgtm. Schip it
